src/sql/SqlService.ts

```diff
@@ -0,0 +1,290 @@
+     * @throws {@link HazelcastSqlException} If there is an error running SQL
+     */
+    execute(sql: SqlStatement): SqlResult;
+}
+
+/** @internal */
+export class SqlServiceImpl implements SqlService {
+    /** Value for the timeout that is not set. */
+    static readonly TIMEOUT_NOT_SET = Long.fromInt(-1);
+    /** Default timeout. */
+    static readonly DEFAULT_TIMEOUT = SqlServiceImpl.TIMEOUT_NOT_SET;
+    /** Default cursor buffer size. */
+    static readonly DEFAULT_CURSOR_BUFFER_SIZE = 4096;
+
+    static readonly DEFAULT_FOR_RETURN_RAW_RESULT = false; // don't return raw result by default
+
+    static readonly DEFAULT_EXPECTED_RESULT_TYPE = SqlExpectedResultType.ANY;
+
+    static readonly DEFAULT_SCHEMA: string | null = null;
+
+    constructor(
+        private readonly connectionRegistry: ConnectionRegistry,
+        private readonly serializationService: SerializationService,
+        private readonly invocationService: InvocationService,
+        private readonly connectionManager: ConnectionManager
+    ) {
+    }
+
+    /**
+     * Handles SQL execute response.
+     * @param clientMessage The response message
+     * @param res SQL result for this response
+     */
+    handleExecuteResponse(clientMessage: ClientMessage, res: SqlResultImpl): void {
+        const response = SqlExecuteCodec.decodeResponse(clientMessage);
+        if (response.error !== null) {
+            res.onExecuteError(
+                new HazelcastSqlException(
+                    response.error.originatingMemberId, response.error.code, response.error.message
+                )
+            );
+        } else {
+            res.onExecuteResponse(
+                response.rowMetadata !== null ? new SqlRowMetadataImpl(response.rowMetadata) : null,
+                response.rowPage,
+                response.updateCount
+            );
+        }
+    }
+
+    /**
+     * Validates SqlStatement
+     *
+     * @param sqlStatement
+     * @throws RangeError if validation is not successful
+     * @internal
+     */
+    static validateSqlStatement(sqlStatement: SqlStatement | null): void {
+        if (sqlStatement === null) {
+            throw new RangeError('Sql statement cannot be null');
+        }
+        tryGetString(sqlStatement.sql);
+        if (sqlStatement.hasOwnProperty('params')) // if params is provided, validate it
+            tryGetArray(sqlStatement.params);
+        if (sqlStatement.hasOwnProperty('options')) // if options is provided, validate it
+            SqlServiceImpl.validateSqlStatementOptions(sqlStatement.options);
+    }
+
+    /**
+     * Validates SqlStatementOptions
+     *
+     * @param sqlStatementOptions
+     * @throws RangeError if validation is not successful
+     * @internal
+     */
+
+    static validateSqlStatementOptions(sqlStatementOptions: SqlStatementOptions): void {
+        if (sqlStatementOptions.hasOwnProperty('schema'))
+            tryGetString(sqlStatementOptions.schema);
```

Comment:
schema can be null on the java side. does this check allow this?

Reply:
I think instead of user giving null, he or she should not give it at all. This way we can simplify validation.

Reply:
Sounds okay

Reply:
Added null support anyway :)